src/util/piped_process.cpp

```diff
@@ -0,0 +1,361 @@
+/// \file piped_process.cpp
+/// Subprocess communication with pipes.
+/// \author Diffblue Ltd.
+
+#ifdef _WIN32
+#  include <tchar.h> // library for _tcscpy function
+#else
+#  include <fcntl.h>  // library for fcntl function
+#  include <poll.h>   // library for poll function
+#  include <signal.h> // library for kill function
+#  include <string.h> // library for strerror function (on linux)
+#  include <unistd.h> // library for read/write/sleep/etc. functions
+#endif
+
+#include <iostream>
+#include <vector>
+
+#include "invariant.h"
+#include "optional.h"
+#include "piped_process.h"
+#include "string_utils.h"
+
+#define BUFSIZE 2048
+
+piped_processt::piped_processt(const std::vector<std::string> commandvec)
+{
+  // Default state
+  process_state = statet::NOT_CREATED;
+#ifdef _WIN32
+  // Security attributes for pipe creation
+  SECURITY_ATTRIBUTES sec_attr;
+  // Ensure pipes are inherited
+  sec_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
+  sec_attr.bInheritHandle = TRUE;
+  sec_attr.lpSecurityDescriptor = NULL;
+  // Use named pipes to allow non-blocking read
+  child_std_IN_Rd = CreateNamedPipe(
+      TEXT("\\\\.\\pipe\\cbmc\\SMT2\\child\\IN"),
+      PIPE_ACCESS_INBOUND,          // Reading for us
```

Comment:
What happens when there are multiple instances of CBMC running simultaneously? Do they all try to talk over the same named pipe?

Reply:
Fixed with use of random number.

Reply:
`rand` isn't random, but produces a deterministic sequence unless you change the seed.
Consider using something you know to be unique, say a process ID.

Reply:
It was seeded, but process ID is better - done.